torchaudio/prototype/io/streamer.py

```diff
@@ -0,0 +1,486 @@
+        an output stream, then the data are discarded.
+
+        When the source reaches EOF, then it triggers all the output stream
+        processors to ender drain mode. All the output stream processors
+        flush the pending frames.
+
+        Returns:
+            int:
+                ``0``
+                A packet was processed properly. The caller can keep
+                calling this function to buffer more frames.
+
+                ``1``
+                The streamer reached EOF. All the output stream processors
+                flushed the pending frames. The caller should stop calling
+                this method.
+
+                ``<0``
+                An error has occured.
+        """
+        return torch.ops.torchaudio.ffmpeg_streamer_process_packet(self._s)
+
+    def process_all_packets(self) -> int:
+        """Keep processing the source until it reaches EOF or an error happens.
+
+        Returns:
+            int:
+                ``1``
+                The streamer reached EOF. All the output stream processors
+                flushed the pending frames.
+
+                ``<0``
+                An error has occured.
+        """
+        return torch.ops.torchaudio.ffmpeg_streamer_process_all_packets(self._s)
+
+    def is_buffer_ready(self) -> bool:
+        """Returns true if all the output streams have at least one chunk filled."""
+        return torch.ops.torchaudio.ffmpeg_streamer_is_buffer_ready(self._s)
+
+    def pop_chunks(self) -> Tuple[torch.Tensor]:
+        """Pop one chunk from all the output stream buffers."""
+        return torch.ops.torchaudio.ffmpeg_streamer_pop_chunks(self._s)
+
+    def fill_buffer(self) -> int:
+        """Keep processing packets until all buffers have at least one chunk
+
+        Returns:
+            int:
+                ``0``
+                Packets are processed properly and buffers are
+                ready to be popped.
+
+                ``1``
+                The streamer reached EOF. All the output stream processors
+                flushed the pending frames. The caller should stop calling
+                this method.
+
+                ``<0``
+                An error has occured.
+        """
+        while not self.is_buffer_ready():
+            for _ in range(3):
+                code = self.process_packet()
+                if code != 0:
+                    return code
+        return 0
+
+    def __iter__(self) -> Iterator[List[torch.Tensor]]:
+        """Return an iterator that generates list of tensors, one per output stream"""
+        while True:
+            code = self.fill_buffer()
+            if code == 0:
+                yield self.pop_chunks()
+            elif code == 1:
+                while True:
+                    chunks = self.pop_chunks()
+                    if all(c is None for c in chunks):
+                        return
+                    yield chunks
+                return
+            else:
+                return
```

Comment:
should the user be alerted that an unexpected error has occurred?

Reply:
Good point and definitely we need to iterate on this but I am not sure what is the best way to communicate the error.

The code that streams [the tutorial MP4](https://download.pytorch.org/torchaudio/tutorial-assets/stream-api/NASAs_Most_Scientifically_Complex_Space_Observatory_Requires_Precision-MP4.mp4) was working fine in NY but now I am facing some issue in Tokyo, maybe because of network (bandwidth?) issue.

```
[tls @ 0x144e041d0] Error in the pull function.
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x147808200] stream 0, offset 0x46f5ac3: partial file
[tls @ 0x1126ccfd0] Error in the pull function.
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x163012000] stream 0, offset 0x59fc887: partial file
[tls @ 0x11262c8c0] Error in the pull function.
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x163012000] stream 0, offset 0x6e951de: partial file
[tls @ 0x12277e830] Error in the pull function.
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x166008600] stream 0, offset 0x37f037f: partial file
[tls @ 0x1126f0ef0] Error in the pull function.
[aac @ 0x16300e200] decode_band_types: Input buffer exhausted before END element found
```

In C++ code, the streaming part does not throw an exception for the sake of performance and recoverability, maybe we should raise an exception on Python layer.

Reply:
@hwangjeff Okay, I changed the semantic on Python layer. Errors that happen during the actual streaming in C++ is notified by return value. The new code convert this into exception so that Python API is more Pythonic.